### DISCUSSION TITLE
[Feature] P3 Thumbnails

### DISCUSSION THREAD

**Discussion by AngelaDMerkel** [immich-app/immich#3623] [Feature Request]
### The feature

Currently Immich generates thumbnails in sRGB which results in the images being notably washed out, especially in the highlights. Most people have devices with support for P3 and offering more options to convert to the colour space in thumbnails would preserve the visual integrity of the display image. Currently sRGB thumbs are notably inferior, especially in the highlights. 

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by AngelaDMerkel** [comment]
This feature can probably be bundled with #3597 and would probably rely on ffmpeg or imagemagick/libvips or whatever is being used to handle the thumbnail conversions.

**Comment by mertalev** [comment]
The issue with HDR images goes beyond whether the display supports them since the browser also needs to support displaying them. IIRC Chrome only recently added support (before that it tone-mapped HDR content to SDR internally), while Firefox doesn't support displaying them at all. 

Then there's the fact that JPEG and WebP are both limited to 8-bit depth, which means we would have to find another format for serving HDR thumbnails.

**Reply by AngelaDMerkel** [reply]
@mertalev P3 is not an HDR colour space and colour space has little relationship to bit depth. There is a colour space called HDR-P3 which has more luminance range, but standard P3 is not HDR and is easily encoded into JPEG. P3 is in incredibly [common use](https://www.wide-gamut.com/test) in the web, and indeed Apple has encoded all their images for years as P3. Every new-ish phone and computer supports it, and most displays cover 99% P3. 

**Emphasis: DCI-P3 is "wide gamut" but not HDR and is happily encoded into 8 bit jpeg. I do it all the time.**

**Reply by mertalev** [reply]
I looked into this a bit more and it seems you're right! I can try adding a setting for wide-gamut.

**Reply by AngelaDMerkel** [reply]
Thank you for the effort. I appreciate you looking into it.

**Comment by AngelaDMerkel** [comment]
Issue #3136 is probably the same as my feature request as the user is using an even wider colour space than P3

**Comment by AngelaDMerkel** [comment]
@mertalev  I appreciated all the time you put into including P3 support. Lightroom has recently introduced full HDR colour space support in the following colour spaces:

- HDR P3
- HDR sRGB
- HDR BT2020 (Rec.2020)

Now that Android 14 is introducing "Ultra HDR" which is just a gain map on top of a standard P3 JPEG it might be worthwhile including HDR image support now too. At this time though there doesn't seem to be an easy way to use ffmpeg or [imagemagick](https://github.com/ImageMagick/ImageMagick/discussions/5441) to do this unfortunately. It always seems to clip the values when converting formats or downscaling in my experience. It would seem to me that the gain map support would be more straightforward to include at least in Android 14. I would be shocked if Apple was fast to adopt

I tried to figure out which colour space Apple is using for their HDR images that the iPhone produces now, but exiftool always reports "uncalibrated" for non-sRGB colour spaces so it's tough for me to figure out. I assume it is HDR P3 because their documentation discusses that as the maximum display support but HDR video in encoded in BT2020 and considering that the images being taken are saved in HEIC it wouldn't surprise me if they were also just BT2020.

**Reply by mertalev** [reply]
Thanks for letting me know about Ultra HDR. That seems like a really nice way to bring HDR to the web. 

For Apple devices, they seem to have something similar called Extended Dynamic Range embedded in HEIC images. I don't know what needs to be done for the image viewer to actually use that metadata, or if it can get picked up automatically.

**Reply by AngelaDMerkel** [reply]
I spent some time looking on this, and at least with older iPhone images HDR display is done in more or less the same way as with Ultra HDR. An additional metadata tag in the images identifies that the image was tone mapped down and guesses how that was done, then applies another tone map to do the reverse and display higher luminance values in parts of the image. 

I don't know what they are doing in newer HDR images, but I can display HDR with them in Photoshop fine, so it's unlikely to be anything proprietary.

**Reply by mertalev** [reply]
It looks like libvips is [looking into this feature](https://github.com/libvips/libvips/issues/3799). If this is added upstream, it would be fairly trivial to support it on our end (at least on web, I'm not sure if the mobile app's photo viewer would use this metadata without other changes).

**Reply by AngelaDMerkel** [reply]
@mertalev I've been following that discussion a while now and think it's likely they implement it, especially in light of the number of HDR and JXL related queries there are. Moreso too that Samsung has implemented HDR JXL. Immich uses vips through sharp correct? If Ultra HDR is supported, will you be willing to add support for HDR P3 and sRGB? Vips already supports those in JXL, EXR etc alongside BT709 and 2020 in AVIF

**Reply by mertalev** [reply]
Definitely, but whether this is default will depend on whether the images still look normal in SDR.

**Reply by AngelaDMerkel** [reply]
My experience and understanding is that they do. Most HDR colour spaces require that you also grade for SDR and fall back to that if HDR is not support. I can provide you a few samples if you're interested. Google Photos already encodes uploaded HDR P3 into UltraHDR and displays them on Samsung, Pixel and desktop Chrome instances.
